SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracer.h"
+
+#include "Misc/ScopeLock.h"
+#include "SpatialGDKSettings.h"
+#include "UObject/Object.h"
+#include "UObject/UnrealType.h"
+#include <WorkerSDK/improbable/c_io.h>
+#include <WorkerSDK/improbable/c_trace.h>
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracer);
+
+using namespace SpatialGDK;
+using namespace worker::c;
+
+void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
+{
+	SpatialEventTracer* EventTracer = static_cast<SpatialEventTracer*>(UserData);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
```

Comment:
`!EventTracer->IsEnabled()` if this is ever true (aka IsEnabled = false) then this is bad memory access, this is only set in the destructor of the class and there should be no callbacks with this userdata beyond that point enforced by the 2 worker API calls here 
```
Trace_EventTracer_Disable(EventTracer);
Trace_EventTracer_Destroy(EventTracer);
```

Think it's safe to remove this and guarantee UserData is valid (and remove confusion around whether it can be valid)

Reply:
Actually, backtracking a bit - if it is enabled and destruction happens after this condition the critical section below will access invalid memory. Might need to synchronize this callback with the EventTracer class externally somehow